[browser][file system] Tests System.IO.FileSystem

kjpou1 opened the pull request:
Added 45 `[PlatformSpecific(~TestPlatforms.Browser)]` checks to tests.

```
 info: test[0]
        Tests run: 3804, Errors: 0, Failures: 0, Skipped: 100. Time: 37.597637s

```

```
// IO.Pipes not supported
public void FlushCanBeUsedOnPipes_Success(bool? flushToDisk)
public void CanSeekReturnsFalseForPipe()
public async Task AnonymousPipeReadViaFileStream(bool asyncReads)
public async Task AnonymousPipeWriteViaFileStream(bool asyncWrites)
public async Task AnonymousPipeViaFileStream_AllDataCopied(int writeSize, int numWrites)

// System.PlatformNotSupportedException : Cannot wait on monitors on this runtime.
public void EmptyFile_Read_Succeeds()
public void NonEmptyFile_Read_GetsExpectedData()
public void NonEmptyWrite_WritesExpectedData()
public Task ManyConcurrentWriteAsyncs()

```

- Directory.CreateDirectory.cs

   - Browser platform volume does not limit each component of the path to a total of 255 characters.
   - Remove the test `DirectoryWithComponentLongerThanMaxComponentAsPath_ThrowsException` for Browser platform.
   - Add new test `DirectoryWithComponentLongerThanMaxComponentAsPath_BrowserDoesNotThrowsException` for Browser platform in case this check is added in the future.


~~Error number:~~  https://github.com/dotnet/runtime/issues/39955 - Resolved

File locking not supported: https://github.com/dotnet/runtime/issues/40065

```
System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

System.IO.FileSystem/tests/File/ReadWriteAllBytesAsync.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public async Task OpenFile_ThrowsIOExceptionAsync()

System.IO.FileSystem/tests/File/ReadWriteAllLines.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

System.IO.FileSystem/tests/File/ReadWriteAllLinesAsync.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public async Task OpenFile_ThrowsIOExceptionAsync()

System.IO.FileSystem/tests/File/ReadWriteAllText.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void OpenFile_ThrowsIOException()

System.IO.FileSystem/tests/File/ReadWriteAllTextAsync.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public async Task OpenFile_ThrowsIOExceptionAsync()

System.IO.FileSystem/tests/File/Create.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void FileInUse()

System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void NoTruncateOnFileShareViolation(FileMode fileMode)

System.IO.FileSystem/tests/FileStream/ctor_str_fm_fa_fs.write.cs
        [ActiveIssue("https://github.com/dotnet/runtime/issues/40065", TestPlatforms.Browser)]
         public void FileShareWithoutWriteThrows()
```

A FIFO special file is similar to a pipe, except that it is created in a different way. Instead of being an anonymous communications channel, a FIFO special file is entered into the file system by calling mkfifo().  Pipes are not supported on WebAssembly.  Platform specific skip for browser in the following tests.

```
System.IO.FileSystem/tests/DirectoryInfo/Exists.cs
        public void FalseForNonRegularFile()

System.IO.FileSystem/tests/Directory/Exists.cs
        public void FalseForNonRegularFile()

System.IO.FileSystem/tests/FileInfo/Exists.cs
        public void TrueForNonRegularFile()

System.IO.FileSystem/tests/File/Exists.cs
        public void FalseForNonRegularFile()
```

~~File rename issue reporting wrong error:~~  https://github.com/dotnet/runtime/issues/40305
**_PR #40310 works around the issue for now while waiting for fix emscripten-core/emscripten#11804 / emscripten-core/emscripten#11812_**

~~Depends on PR: https://github.com/dotnet/runtime/pull/39763~~ - merged

kjpou1 committed:
[browser][file system] Tests System.IO.FileSystem

steveisok commented on src/libraries/System.IO.FileSystem/tests/FileSystemTest.Browser.cs:
```diff
@@ -0,0 +1,21 @@
+    {
+        protected static int geteuid()
+        {
+            throw new PlatformNotSupportedException();
```
What do you think about returning 0 here and eliminating the platform checks?

kjpou1 replied:
By doing that and modifying the places that use `geteuid` the tests fail with:

```
[FAIL] System.IO.Tests.File_AppendAllLinesAsync.WriteToReadOnlyFile
        
  [40m[37mdbug[39m[22m[49m: test[0]
        System.UnauthorizedAccessException : Access to the path '/tmp/File_AppendAllLinesAsync_mqajhnoo.uqp/WriteToReadOnlyFile_105_93d848c1' is denied.
        
  [40m[37mdbug[39m[22m[49m: test[0]
        ---- System.IO.IOException : Permission denied
```

steveisok replied:
The pal uses `sendFile` and emscripten's implementation doesn't seem too friendly w/ read only files.  The error message refers to not being able to write to the destination file.

If you skip this block https://github.com/dotnet/runtime/blob/05f17e3170d014efe63202c4e7e7b641e8f8fef6/src/libraries/Native/Unix/System.Native/pal_io.c#L1152-L1191, I think copying will work.

kjpou1 replied:
I have gone through the build and it looks like this code block is already being skipped in the native pal.

steveisok replied:
You're right.  I was looking at the wrong place thinking it was included.  

We should trace to see what call is bad.

kjpou1 replied:
agreed

MaximLipnin commented:
You also need to remove the test suite from `src/libraries/tests.proj`:
https://github.com/dotnet/runtime/blob/master/src/libraries/tests.proj#L32

jkotas commented on src/libraries/System.IO.FileSystem/tests/Directory/Move.cs:
```diff
@@ -33,6 +33,7 @@ public void EmptyPath()
         }
 
         [Fact]
+        [PlatformSpecific(~TestPlatforms.Browser)]
```
Is this a bug, or does this really exercise a functionality not supported in browser?

kjpou1 replied:
Still need to go through these and make sure they really are not supported or bugs or whatever.

steveisok replied:
The VFS does have odd behavior for some of these tests.  We can likely work around some of that in the pal, but this suite in particular is going to need a bit of analysis like Ken was saying.

stephentoub replied:
We should have an issue tracking it then and use ActiveIssue instead. Thanks.

kjpou1 replied:
What is the preferred way to comment on a platform specific test.  For instance the `public void FlushCanBeUsedOnPipes_Success(bool? flushToDisk)` test run into PNSE IO.Pipes.  Should this use an ActiveIssue?  Does not make sense to me for that type of case but want to make sure each case is documented for the PR.

kjpou1 replied:
Is just a comment in the code sufficient for instance?  `[PlatformSpecific(~TestPlatforms.Browser)] // IO.Pipes not supported`

jkotas replied:
If the problem is understood and by design, comment like what you have suggested looks good to me.

jkotas commented on src/libraries/System.IO.FileSystem/tests/DirectoryInfo/MoveTo.cs:
```diff
@@ -24,6 +24,7 @@ protected virtual void Move(DirectoryInfo sourceDir, string destDir)
         #region UniversalTests
 
         [Fact]
+        [PlatformSpecific(~TestPlatforms.Browser)]
```
Same question here

kjpou1 commented:
> You also need to remove the test suite from `src/libraries/tests.proj`:
> https://github.com/dotnet/runtime/blob/master/src/libraries/tests.proj#L32

Thanks Maxim was just waiting for PR https://github.com/dotnet/runtime/pull/39763 because it would have failed anyway

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Remove FileSystem from test project exclusions.

mdh1418 commented on src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -106,7 +107,7 @@ public void WriteToReadOnlyFile()
             {
                 // Operation succeeds when being run by the Unix superuser
-                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && geteuid() == 0)
+                if ((!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && !RuntimeInformation.IsOSPlatform(OSPlatform.Browser))  && geteuid() == 0)
```
Remove extra set of parenthesis with short circuiting here as well as in the five files below?

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
[browser][filesystem] Comment on skipped tests due to IO.Pipes not supported

kjpou1 committed:
[browser][filesystem] Comment on skipped tests due to monitor not supported

kjpou1 committed:
Address review comments

MaximLipnin commented on src/libraries/System.IO.FileSystem/tests/FileStream/ReadWriteSpan.cs:
```diff
@@ -24,6 +24,7 @@ public void DisposedStream_ReadWrite_Throws()
         }
 
         [Fact]
+        [PlatformSpecific(~TestPlatforms.Browser)] // PNSE: Cannot wait on monitors
```
Don't we use `[ConditionalFact(typeof(PlatformDetection), nameof(PlatformDetection.IsThreadingSupported))]` for such a case?

kjpou1 replied:
Addressed.  Thanks Maxim

kjpou1 committed:
Address review comments on how to handle monitor PNSE

jkotas commented on src/libraries/System.IO.FileSystem/tests/File/GetSetTimes.cs:
```diff
@@ -13,7 +13,7 @@ public class File_GetSetTimes : StaticGetSetTimes
         // OSX has the limitation of setting upto 2262-04-11T23:47:16 (long.Max) date.
         // 32bit Unix has time_t up to ~ 2038.
-        private static bool SupportsLongMaxDateTime => PlatformDetection.IsWindows || (RuntimeInformation.ProcessArchitecture != Architecture.Arm && RuntimeInformation.ProcessArchitecture != Architecture.X86 && !PlatformDetection.IsOSXLike);
+        private static bool SupportsLongMaxDateTime => PlatformDetection.IsWindows || (RuntimeInformation.ProcessArchitecture != Architecture.Arm && RuntimeInformation.ProcessArchitecture != Architecture.X86 && !PlatformDetection.IsOSXLike && !PlatformDetection.IsBrowser);
```
```suggestion
        private static bool SupportsLongMaxDateTime => PlatformDetection.IsWindows || (!PlatformDetection.Is32BitProcess && !PlatformDetection.IsOSXLike);
```

jkotas commented on src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -106,7 +107,7 @@ public void WriteToReadOnlyFile()
             {
                 // Operation succeeds when being run by the Unix superuser
-                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && geteuid() == 0)
+                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && !RuntimeInformation.IsOSPlatform(OSPlatform.Browser)  && geteuid() == 0)
```
```suggestion
                if (PlatformDetection.IsSuperUser)
```

jkotas replied:
(multiple places)

kjpou1 replied:
Thanks for the IsSuperUser hint.  Did not know about that.

kjpou1 replied:
Addressed in recent commits

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Update src/libraries/System.IO.FileSystem/tests/File/GetSetTimes.cs

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

kjpou1 committed:
Merge branch 'wasm-filesystem-tests' of https://github.com/kjpou1/runtime into wasm-filesystem-tests

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Add ActiveIssue for browser platform

```
        [ActiveIssue("https://github.com/dotnet/runtime/issues/39955", TestPlatforms.Browser)]
        public void NotFoundErrorIsExpected()
```

kjpou1 committed:
Simplify test case with `PlatformDetection.IsSuperUser` as per review comment

kjpou1 committed:
Remove unused variable from tests

kjpou1 committed:
Browser platform volume does not limit each component of the path to a total of 255 characters.

- Remove the test `DirectoryWithComponentLongerThanMaxComponentAsPath_ThrowsException` for Browser platform.
- Add new test `DirectoryWithComponentLongerThanMaxComponentAsPath_BrowserDoesNotThrowsException` for Browser platform in case this check is added in the future.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Fix windows tests

jkotas commented on src/libraries/System.IO.FileSystem/tests/File/ReadWriteAllBytes.cs:
```diff
@@ -106,7 +107,7 @@ public void WriteToReadOnlyFile()
             {
                 // Operation succeeds when being run by the Unix superuser
-                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && geteuid() == 0)
+                if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && PlatformDetection.IsSuperUser)
```
I think it would be better to make `IsSuperUser` return false on Windows instead of throwing.

kjpou1 replied:
Addressed in recent commits

kjpou1 committed:
Add ActiveIssue to test System.IO.Tests.DirectoryInfo_Name.CurrentDirectory

- System.IO.Tests.DirectoryInfo_Name.CurrentDirectory does not pass because it is using Path.GetFileName on the current directory of "/".
- See issue https://github.com/dotnet/runtime/issues/39998 for more information

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Add active issue for tests that need to support file locking.

kjpou1 committed:
Add active issue for tests that need to support file locking.

kjpou1 committed:
Browser volume does not have a limit on segments

kjpou1 committed:
Remove browser platform test

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Remove active issue for NotFoundErrorIsExpected

kjpou1 committed:
Remove platform specific from SkippingHiddenFiles

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Fix enumeration test errors failing from `HiddenFilesAreReturned`.

- WebAssembly (BROWSER) has dirent d_type but is not identifying correctly but by returning UNKNOWN the managed code properly stats the file to detect if entry is directory or not.

kjpou1 committed:
Fix enumeration test errors failing from `HiddenFilesAreReturned`.

- WebAssembly (BROWSER) has dirent d_type but is not identifying correctly but by returning UNKNOWN the managed code properly stats the file to detect if entry is directory or not.

kjpou1 committed:
Fix build for TARGET_WASM not defined

kjpou1 committed:
Address review comment by add check for IsWindows to IsSuperUser.

kjpou1 committed:
Address review comment by add check for IsWindows to IsSuperUser.

kjpou1 committed:
Platform Specific test for hidden files no longer needed.

kjpou1 committed:
Platform Specific test fno longer needed.

mdh1418 commented:
There were a few places in `System.IO.MemoryMappedFiles` that also had `(!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && PlatformDetection.IsSuperUser)`, since we are changing `IsSuperUser` to return false rather than throwing here, should those two instances in [`MemoryMappedFiles.CreateFromFile.Tests.cs`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs) [L700](https://github.com/dotnet/runtime/blob/215db60d405f6a963121b54035a35163681ff5b7/src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs#L700) and [L706](https://github.com/dotnet/runtime/blob/215db60d405f6a963121b54035a35163681ff5b7/src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs#L706) be changed in this PR as well? If we don't want to mix suites, I can just put up another PR after this gets merged.

kjpou1 commented:
@mdh1418

> There were a few places in `System.IO.MemoryMappedFiles` that also had `(!RuntimeInformation.IsOSPlatform(OSPlatform.Windows) && PlatformDetection.IsSuperUser)`, since we are changing `IsSuperUser` to return false rather than throwing here, should those two instances in [`MemoryMappedFiles.CreateFromFile.Tests.cs`](https://github.com/dotnet/runtime/blob/master/src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs) [L700](https://github.com/dotnet/runtime/blob/215db60d405f6a963121b54035a35163681ff5b7/src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs#L700) and [L706](https://github.com/dotnet/runtime/blob/215db60d405f6a963121b54035a35163681ff5b7/src/libraries/System.IO.MemoryMappedFiles/tests/MemoryMappedFile.CreateFromFile.Tests.cs#L706) be changed in this PR as well? If we don't want to mix suites, I can just put up another PR after this gets merged.

Could you check the files that are in the PR.  I thought I got all of them.

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
Use active issue for rename issue

kjpou1 committed:
Merge branch 'master' of https://github.com/dotnet/runtime into wasm-filesystem-tests

kjpou1 committed:
ActiveIssue no longer needed

- PR https://github.com/dotnet/runtime/pull/40310 works around the issue for now while waiting for fix https://github.com/emscripten-core/emscripten/issues/11804 / https://github.com/emscripten-core/emscripten/pull/11812

Commit:
Use ActiveIssue for SettingUpdatesProperties

Commit:
Use ActiveIssue for TimesIncludeMillisecondPart

Commit:
Use ActiveIssue for ErrorHandlingTests failures

Commit:
Use ActiveIssue for SetUptoNanoseconds

Commit:
Use ActiveIssue for GetSetTimes test failures

Commit:
Use ActiveIssue for tests failing with DirectoryNotFoundException

Commit:
Use ActiveIssue for tests failing with UnauthorizedAccessException

mdh1418 commented:
Currently the results are: `Tests run: 3903, Errors: 0, Failures: 0, Skipped: 117. Time: 10.340306s`

MaximLipnin commented on src/libraries/Native/Unix/System.Native/pal_io.c:
```diff
@@ -259,6 +262,7 @@ static int32_t ConvertOpenFlags(int32_t flags)
 
 intptr_t SystemNative_Open(const char* path, int32_t flags, int32_t mode)
 {
+    //fprintf(stderr, "SystemNative_Open %s flags: %i  mode: %i\n ", path, flags, mode);
```
Is it still required?

mdh1418 replied:
Thanks for the catch

MaximLipnin commented on src/libraries/System.IO.FileSystem/tests/Directory/CreateDirectory.cs:
```diff
@@ -210,17 +210,26 @@ public void DirectoryEqualToMaxDirectory_CanBeCreatedAllAtOnce()
+        [Theory,
+            MemberData(nameof(PathsWithComponentLongerThanMaxComponent))]
+        [PlatformSpecific(TestPlatforms.Browser)] // Browser specific test in case the check changes in the future
+        public void DirectoryWithComponentLongerThanMaxComponentAsPath_BrowserDoesNotThrowsException(string path)
```
typo: DirectoryWithComponentLongerThanMaxComponentAsPath_BrowserDoesNotThrow**s**Exception(string path)

mdh1418 replied:
Thanks :)

MaximLipnin reviewed (approved):
lgtm, just a couple of minor comments

Commit:
Remove debugging statement from pal_io

Commit:
Address type in CreateDirectory method name

Commit:
Merge remote-tracking branch 'upstream/master' into wasm-filesystem-tests
